src/python/pants/jvm/compile.py

```diff
@@ -89,10 +128,21 @@ def for_targets(
         request types which are marked `root_only`.
         """
 
+        for (input, request_types) in jvm_request_types.code_generator_requests.items():
+            if not component.representative.has_field(input):
+                continue
+            if len(request_types) > 1:
+                raise ClasspathSourceAmbiguity(
```

Comment:
at the moment, this will result in a late failure, and only if a user attempts to use a source file with conflicting implementations as a dependency. I could re-write to fail fast, but the code will eventually need to fail here once #14041 is in place.

Reply:
It's still not clear to me why this matching isn't happening inside of `ClasspathEntryRequest.classify_impl`. If it did, both 1) only matching the representative, 2) late erroring... would be handled by the existing code below, right? Because you'd match multiple impls.